mlir/lib/Dialect/Rock/IR/GetRockInfo.cpp

```
@@ -0,0 +1,136 @@
+      return features.getValue();
+
+    // If the initial op is a func and there is no `features` attribute, then
+    // we cannot proceed
+    if (isa<func::FuncOp>(op) || isa<gpu::GPUFuncOp>(op))
+      llvm_unreachable("Trying to get 'features' for an invalid func op");
+  }
+
+  // Next, check to see if the op has a 'features' attribute.
+  if (auto features = op->getAttrOfType<rock::GemmFeaturesAttr>("features"))
+    return features.getValue();
+
+  // In this case, the op does not have a 'Features' attribute, so we can
+  // calculate the default features based on the architecture.
+  rock::AmdArchInfo archInfo = rock::lookupArchInfo(rock::getArchValue(op));
+  // Get the types needed for feature calculation using TypeSwitch
+  SmallVector<Type> typesForFeature =
+      llvm::TypeSwitch<Operation *, SmallVector<Type>>(op)
+          .Case<rock::GridwiseGemmOp, rock::GridwiseGemmAccelOp,
+                rock::BlockwiseGemmAccelOp, rock::ThreadwiseAccelGemmOp,
+                rock::GridwiseAttentionAccelOp, RockGemmGemmWrapperInterface,
+                RockGemmWrapperInterface>([](auto opWithFeatures) {
+            return opWithFeatures.getTypesForFeature();
+          })
+          .Default([](Operation *op) -> SmallVector<Type> {
+            llvm_unreachable("Trying to get feature type on unsupported op");
+          });
+
+  std::optional<rock::GemmFeatures> features = std::nullopt;
+  for (auto &ty : typesForFeature) {
+    // If features is not yet set, then we can update features without having to
+    // do an set intersection first
+    auto newFeatures = archInfo.getDefaultFeatures(ty);
+    if (!features.has_value()) {
+      features = newFeatures;
+      continue;
+    }
+
+    // For all other types, we need to do a set intersection
+    features = intersectGemmFeatures(features.value(), newFeatures);
+  }
+
+  // Handle the case where no types were found, and we could not calculate
+  // features
+  if (!features.has_value()) {
```

this should only happen if typesForFeature.size() == 0? I think we should fail in this case? @umangyadav

---

Yeah, this can only happen for typesForFeature.size() == 0. I've updated the logic to fail instead of setting `GemmFeatures::none`